### PR TITLE
Add, update, reorganize Constants

### PR DESCRIPTION
**Pull request by DonovanDMC**
This converts several unions into constants, and converts several hardcoded numbers and such into constant references (haven't searched through the js), as well as introduces several parity fixes for staying as close to Discord as reasonably possible.

### Added Constants
* [`ConnectionVisibilityTypes`](https://discord.com/developers/docs/resources/user#connection-object-visibility-types)
* [`DefaultMessageNotificationLevels`](https://discord.com/developers/docs/resources/guild#guild-object-default-message-notification-level)
* [`ExplicitContentFilterLevels`](https://discord.com/developers/docs/resources/guild#guild-object-explicit-content-filter-level)
* [`GuildIntegrationExpireBehavior`](https://discord.com/developers/docs/resources/guild#integration-object-integration-expire-behaviors)
* [`GuildIntegrationTypes`](https://discord.com/developers/docs/resources/guild#integration-object-integration-structure)
* [`GuildNSFWLevels`](https://discord.com/developers/docs/resources/guild#guild-object-guild-nsfw-level)
* [`InviteTargetTypes`](https://discord.com/developers/docs/resources/invite#invite-object-invite-target-types)
* [`MFALevels`](https://discord.com/developers/docs/resources/guild#guild-object-mfa-level)
* [`MessageActivityTypes`](https://discord.com/developers/docs/resources/channel#message-object-message-activity-types)
* [`PermissionOverwriteTypes`](https://discord.com/developers/docs/resources/channel#overwrite-object)
* [`PremiumTiers`](https://discord.com/developers/docs/resources/guild#guild-object-premium-tier)
* [`PremiumTypes`](https://discord.com/developers/docs/resources/user#user-object-premium-types)
* [`StageInstancePrivacyLevel`](https://discord.com/developers/docs/resources/stage-instance#stage-instance-object-privacy-level)
* [`SystemChannelFlags`](https://discord.com/developers/docs/resources/guild#guild-object-system-channel-flags)
* [`VerificationLevels`](https://discord.com/developers/docs/resources/guild#guild-object-verification-level)
* [`VideoQualityModes`](https://discord.com/developers/docs/resources/channel#channel-object-video-quality-modes)
* `ThreadMemberFlags` ([not documented](https://butts-are.cool/notepad__11-14-2021_02-14-54.png))
* [`WebhookTypes`](https://discord.com/developers/docs/resources/webhook#webhook-object-webhook-types)

##### [ChannelTypes](https://discord.com/developers/docs/resources/channel#channel-object-channel-types)
* `GUILD_STAGE` -> `GUILD_STAGE_VOICE`

##### [GatewayOPCodes](https://discord.com/developers/docs/topics/opcodes-and-status-codes#gateway-gateway-opcodes)
* `EVENT` -> `DISPATCH`
* `STATUS_UPDATE` -> `PRESENCE_UPDATE`
* `GET_GUILD_MEMBERS` -> `REQUEST_GUILD_MEMBERS`

##### [Permissions](https://discord.com/developers/docs/topics/permissions#permissions-bitwise-permission-flags)
* Added  `manageEvents`
* Added  `startEmbeddedActivities`

##### [UserFlags](https://discord.com/developers/docs/resources/user#user-object-user-flags)
* (`OFFICIAL_NAME` (`OUR_NAME`))
* Added `DISCORD_STAFF` (`DISCORD_EMPLOYEE`)
* Added `PARTNER` (`PARTNERED_SERVER_OWNER`
* Added `HYPESQUAD` (`HYPESQUAD_EVENTS`)
* Added `HYPESQUAD_ONLINE_HOUSE_1` (`HOUSE_BRAVERY`)
* Added `HYPESQUAD_ONLINE_HOUSE_2` (`HOUSE_BRILLIANCE`)
* Added `HYPESQUAD_ONLINE_HOUSE_3` (`HOUSE_BALANCE`)
* Added `PREMIUM_EARLY_SUPPORTER` (`EARLY_SUPPORTER`)
* Added `TEAM_PSEUDO_USER` (`TEAM_USER`)
* Added `VERIFIED_DEVELOPER` (`EARLY_VERIFIED_BOT_DEVELOPER`)
* Added `CERTIFIED_MODERATOR` (`DISCORD_CERTIFIED_MODERATOR`)
* Added  `BOT_HTTP_INTERACTIONS`

##### [VoiceOPCodes](https://discord.com/developers/docs/topics/opcodes-and-status-codes#voice-voice-opcodes)
* `DISCONNECT` -> `CLIENT_DISCONNECT`

##### Additional Changes
* both Constant types & js have been re-alphabetized

### CODE REVIEW

**Review by bsian03 — CHANGES_REQUESTED** [review]
What’s with all the new lines you added everywhere in typings

**Comment by HeadTriXz** [comment]
> What’s with all the new lines you added everywhere in typings 

Most likely eslint's auto-fix.

**Comment by DonovanDMC** [comment]
Our linter's auto fix sucks and does that, I didn't add those

**Comment by DonovanDMC** [comment]
the hell did I do

**Comment by DonovanDMC** [comment]
there, god

File: index.d.ts [hunk 1]
```diff
@@ -21,13 +21,18 @@ declare namespace Eris {
...
   type AnyGuildChannel = GuildTextableChannel | AnyVoiceChannel | CategoryChannel | StoreChannel;
   type AnyVoiceChannel = VoiceChannel | StageChannel;
-  type ChannelTypes = Constants["ChannelTypes"][keyof Constants["ChannelTypes"]];
   type GuildTextableChannel = TextChannel | NewsChannel;
   type InviteChannel = InvitePartialChannel | Exclude<AnyGuildChannel, CategoryChannel>;
   type PossiblyUncachedTextable = Textable | Uncached;
   type PossiblyUncachedTextableChannel = TextableChannel | Uncached;
   type TextableChannel = (GuildTextable & GuildTextableChannel) | (Textable & PrivateChannel);
-  type VideoQualityMode = 1 | 2;
+  type VideoQualityMode = (Constants["VideoQualityMode"])[keyof Constants["VideoQualityMode"]];
```
Comment (bsian03) [inline index.d.ts]:
```suggestion
  type VideoQualityMode = Constants["VideoQualityMode"][keyof Constants["VideoQualityMode"]];
```

The () brackets are useless here

File: index.d.ts [hunk 2]
```diff
@@ -21,13 +21,18 @@ declare namespace Eris {
...
+  type ChannelTypes = GuildChannelTypes | PrivateChannelTypes;
+  type GuildChannelTypes = Exclude<(Constants["ChannelTypes"])[keyof Constants["ChannelTypes"]], PrivateChannelTypes>;
```
Comment (bsian03) [inline index.d.ts]:
```suggestion
  type GuildChannelTypes = Exclude<Constants["ChannelTypes"][keyof Constants["ChannelTypes"]], PrivateChannelTypes>;
```
And here

File: index.d.ts [hunk 3]
```diff
@@ -21,13 +21,18 @@ declare namespace Eris {
...
+  type TextChannelTypes = GuildTextChannelTypes | PrivateChannelTypes;
+  type GuildTextChannelTypes = (Constants["ChannelTypes"])[keyof Pick<Constants["ChannelTypes"], "GUILD_TEXT" | "GUILD_NEWS">];
+  type GuildVoiceChannelTypes = (Constants["ChannelTypes"])[keyof Pick<Constants["ChannelTypes"], "GUILD_VOICE" | "GUILD_STAGE">];
+  type PrivateChannelTypes = (Constants["ChannelTypes"])[keyof Pick<Constants["ChannelTypes"], "DM" | "GROUP_DM">];
```
Comment (bsian03) [inline index.d.ts]:
```suggestion
  type GuildTextChannelTypes = Constants["ChannelTypes"][keyof Pick<Constants["ChannelTypes"], "GUILD_TEXT" | "GUILD_NEWS">];
  type GuildVoiceChannelTypes = Constants["ChannelTypes"][keyof Pick<Constants["ChannelTypes"], "GUILD_VOICE" | "GUILD_STAGE">];
  type PrivateChannelTypes = Constants["ChannelTypes"][keyof Pick<Constants["ChannelTypes"], "DM" | "GROUP_DM">];
```
And all of here

File: index.d.ts [hunk 4]
```diff
@@ -44,13 +49,14 @@ declare namespace Eris {
...
 
   // Guild
-  type DefaultNotifications = 0 | 1;
-  type ExplicitContentFilter = 0 | 1 | 2;
-  type GuildFeatures = "ANIMATED_ICON" | "BANNER" | "COMMERCE" | "COMMUNITY" | "DISCOVERABLE" | "FEATURABLE" | "INVITE_SPLASH" | "MEMBER_VERIFICATION_GATE_ENABLED" | "NEWS" | "PARTNERED" | "PREVIEW_ENABLED" | "VANITY_URL" | "VERIFIED" | "VIP_REGIONS" | "WELCOME_SCREEN_ENABLED" | "TICKETED_EVENTS_ENABLED" | "MONETIZATION_ENABLED" | "MORE_STICKERS" | "THREE_DAY_THREAD_ARCHIVE" | "SEVEN_DAY_THREAD_ARCHIVE" | "PRIVATE_THREADS";
-  type NSFWLevel = 0 | 1 | 2 | 3;
+  type DefaultNotifications = (Constants["DefaultMessageNotificationLevel"])[keyof Constants["DefaultMessageNotificationLevel"]];
+  type ExplicitContentFilter = (Constants["ExplicitContentFilterLevel"])[keyof Constants["ExplicitContentFilterLevel"]];
+  type GuildFeatures = Constants["GuildFeatures"][number];
+  type NSFWLevel = (Constants["GuildNSFWLevel"])[keyof Constants["GuildNSFWLevel"]];
```
Comment (bsian03) [inline index.d.ts]:
```suggestion
  type DefaultNotifications = Constants["DefaultMessageNotificationLevel"][keyof Constants["DefaultMessageNotificationLevel"]];
  type ExplicitContentFilter = Constants["ExplicitContentFilterLevel"][keyof Constants["ExplicitContentFilterLevel"]];
  type GuildFeatures = Constants["GuildFeatures"][number];
  type NSFWLevel = Constants["GuildNSFWLevel"][keyof Constants["GuildNSFWLevel"]];
```
Here too

File: index.d.ts [hunk 5]
```diff
@@ -44,13 +49,14 @@ declare namespace Eris {
...
   type PossiblyUncachedGuild = Guild | Uncached;
-  type PremiumTier = 0 | 1 | 2 | 3;
-  type VerificationLevel = 0 | 1 | 2 | 3 | 4;
+  type PremiumTier = (Constants["PremiumTier"])[keyof Constants["PremiumTier"]];
+  type VerificationLevel = (Constants["VerificationLevel"])[keyof Constants["VerificationLevel"]];
+  type SystemChannelFlags = (Constants["SystemChannelFlags"])[keyof Constants["SystemChannelFlags"]];
```
Comment (bsian03) [inline index.d.ts]:
```suggestion
  type PremiumTier = Constants["PremiumTier"][keyof Constants["PremiumTier"]];
  type VerificationLevel = Constants["VerificationLevel"][keyof Constants["VerificationLevel"]];
  type SystemChannelFlags = Constants["SystemChannelFlags"][keyof Constants["SystemChannelFlags"]];
```
Almost forgot about here

File: index.d.ts [hunk 6]
```diff
@@ -69,18 +75,18 @@ declare namespace Eris {
   type ActionRowComponents = Button | SelectMenu;
   type Button = InteractionButton | URLButton;
   type Component = ActionRow | ActionRowComponents;
-  type ImageFormat = "jpg" | "jpeg" | "png" | "gif" | "webp";
+  type ImageFormat = Constants["ImageFormats"][number];
   type MessageContent = string | AdvancedMessageContent;
-  type MFALevel = 0 | 1;
+  type MFALevel = (Constants["MFALevel"])[keyof Constants["MFALevel"]];
```
Comment (bsian03) [inline index.d.ts]:
```suggestion
  type MFALevel = Constants["MFALevel"][keyof Constants["MFALevel"]];
```
You guessed it, here included

File: index.d.ts [hunk 7]
```diff
@@ -69,18 +75,18 @@ declare namespace Eris {
...
   type PossiblyUncachedMessage = Message | { channel: TextableChannel | { id: string; guild?: Uncached }; guildID?: string; id: string };
-  type InteractionType = 1 | 2;
+  type InteractionType = (Constants["ApplicationCommandTypes"])[keyof Constants["ApplicationCommandTypes"]];
```
Comment (bsian03) [inline index.d.ts]:
```suggestion
  type InteractionType = Constants["ApplicationCommandTypes"][keyof Constants["ApplicationCommandTypes"]];
```
Hear here

File: index.d.ts [hunk 8]
```diff
@@ -69,18 +75,18 @@ declare namespace Eris {
...
 
   // Permission
-  type PermissionType = 0 | 1;
+  type PermissionType = (Constants["PermissionOverwriteType"])[keyof Constants["PermissionOverwriteType"]];
```
Comment (bsian03) [inline index.d.ts]:
```suggestion
  type PermissionType = Constants["PermissionOverwriteType"][keyof Constants["PermissionOverwriteType"]];
```
mmm

File: index.d.ts [hunk 9]
```diff
@@ -69,18 +75,18 @@ declare namespace Eris {
...
 
   // Presence/Relationship
-  type ActivityType = BotActivityType | 4;
-  type BotActivityType = 0 | 1 | 2 | 3 | 5;
+  type ActivityType = BotActivityType | Constants["ActivityType"]["CUSTOM"];
+  type BotActivityType = (Constants["ActivityType"])[Exclude<keyof Constants["ActivityType"], "CUSTOM">];
```
Comment (bsian03) [inline index.d.ts]:
```suggestion
  type BotActivityType = Constants["ActivityType"][Exclude<keyof Constants["ActivityType"], "CUSTOM">];
```
Here it ends

File: index.d.ts [hunk 10]
```diff
@@ -511,8 +518,8 @@ declare namespace Eris {
     channelPinUpdate: [channel: TextableChannel, timestamp: number, oldTimestamp: number];
     channelRecipientAdd: [channel: GroupChannel, user: User];
     channelRecipientRemove: [channel: GroupChannel, user: User];
-    channelUpdate: [channel: AnyGuildChannel, oldChannel: OldGuildChannel | OldGuildTextChannel | OldGuildVoiceChannel] 
-      | [channel: GroupChannel, oldChannel: OldGroupChannel];
+    channelUpdate: [channel: AnyGuildChannel, oldChannel: OldGuildChannel | OldGuildTextChannel | OldGuildVoiceChannel]
+    | [channel: GroupChannel, oldChannel: OldGroupChannel];
```
Comment (bsian03) [inline index.d.ts]:
```suggestion
      | [channel: GroupChannel, oldChannel: OldGroupChannel];
```

File: index.d.ts [hunk 11]
```diff
@@ -552,8 +559,8 @@ declare namespace Eris {
     relationshipAdd: [relationship: Relationship];
     relationshipRemove: [relationship: Relationship];
     relationshipUpdate: [relationship: Relationship, oldRelationship: { type: number }];
-    typingStart: [channel: GuildTextableChannel | Uncached, user: User | Uncached, member: Member] 
-      | [channel: PrivateChannel | Uncached, user: User | Uncached, member: null];
+    typingStart: [channel: GuildTextableChannel | Uncached, user: User | Uncached, member: Member]
+    | [channel: PrivateChannel | Uncached, user: User | Uncached, member: null];
```
Comment (bsian03) [inline index.d.ts]:
```suggestion
      | [channel: PrivateChannel | Uncached, user: User | Uncached, member: null];
```

File: index.d.ts [hunk 12]
```diff
@@ -1214,6 +1222,7 @@ declare namespace Eris {
       SYNC_CALL: 13;
     };
     ImageFormats: ["jpg", "jpeg", "png", "webp", "gif"];
+    GuildFeatures: ["ANIMATED_ICON" | "BANNER" | "COMMERCE" | "COMMUNITY" | "DISCOVERABLE" | "FEATURABLE" | "INVITE_SPLASH" | "MEMBER_VERIFICATION_GATE_ENABLED" | "NEWS" | "PARTNERED" | "PREVIEW_ENABLED" | "VANITY_URL" | "VERIFIED" | "VIP_REGIONS" | "WELCOME_SCREEN_ENABLED" | "TICKETED_EVENTS_ENABLED" | "MONETIZATION_ENABLED" | "MORE_STICKERS" | "THREE_DAY_THREAD_ARCHIVE" | "SEVEN_DAY_THREAD_ARCHIVE" | "PRIVATE_THREADS"];
```
Comment (bsian03) [inline index.d.ts]: This means the array only contains 1 element which could be any of those values

File: index.d.ts [hunk 13]
```diff
@@ -1381,6 +1390,81 @@ declare namespace Eris {
       RESUMED: 9;
       DISCONNECT: 13;
     };
+    ApplicationCommandTypes: {
```
Comment (bsian03) [inline index.d.ts]: Alphabetical order pls

File: index.d.ts [hunk 14]
```diff
@@ -1761,6 +1845,8 @@ declare namespace Eris {
       messageID: string,
       options: MessageWebhookContent
     ): Promise<Message<GuildTextableChannel>>;
+    emit<K extends keyof ClientEvents>(event: K, ...args: ClientEvents[K]): boolean;
+    emit(event: string, ...args: any[]): boolean;
```
Comment (bsian03) [inline index.d.ts]: Event stuff should go at the bottom, if linter doesn't allow it, the config should be changed instead

File: index.d.ts [hunk 15]
```diff
@@ -2004,7 +2088,7 @@ declare namespace Eris {
   export class ExtendedUser extends User {
     email: string;
     mfaEnabled: boolean;
-    premiumType: 0 | 1 | 2;
+    premiumType: (Constants["PremiumType"])[keyof Constants["PremiumType"]];
```
Comment (bsian03) [inline index.d.ts]:
```suggestion
    premiumType: Constants["PremiumType"][keyof Constants["PremiumType"]];
```
Oh man there's more here

**Review by bsian03 — DISMISSED** [review]
noob

File: index.d.ts (hunk 2 above)
Comment (DonovanDMC) [reply index.d.ts]: I did that on purpose, you're the noob
but fine, I'll remove them
hmph

File: index.d.ts (hunk 13 above)
Comment (DonovanDMC) [reply index.d.ts]: reeee

File: index.d.ts (hunk 14 above)
Comment (DonovanDMC) [reply index.d.ts]: linter moves it around, alphabetical ordering

File: index.d.ts (hunk 15 above)
Comment (DonovanDMC) [reply index.d.ts]: god damn it bsian

File: index.d.ts (hunk 13 above)
Comment (DonovanDMC) [reply index.d.ts]: we need a rule for this, current ones don't cover this

File: index.d.ts (hunk 12 above)
Comment (DonovanDMC) [reply index.d.ts]: forgot to change the pipes to commas

File: index.d.ts [hunk 16]
```diff
@@ -896,9 +903,10 @@ declare namespace Eris {
     before?: string;
     limit?: number;
   }
+
   interface InteractionButton extends ButtonBase {
     custom_id: string;
-    style: 1 | 2 | 3 | 4;
+    style: (Constants["ButtonType"])[Exclude<keyof Constants["ButtonType"], "LINK">];
```
Comment (coolcalcacol) [inline index.d.ts:909]: >:)

Comment (coolcalcacol) [inline index.d.ts]: Noob.

File: index.d.ts (hunk 13 above)
Comment (DonovanDMC) [reply index.d.ts]: done

**Comment by DonovanDMC** [comment]
Converted to draft at the request of bsian, this will be marked as ready when #1216, and any other major things that should be added to the constants get merged into dev.

**Comment by DonovanDMC** [comment]
This should be all ready to go, though I haven't done thorough testing

**Review by abalabahaha — DISMISSED** [review]

**Review by abalabahaha — APPROVED** [review]

**Pull request merged**